### PR TITLE
Add config module as per #2349

### PR DESCRIPTION
**Pull request by belluzj**
See issue #2349

### CODE REVIEW

**Comment by belluzj** [comment]
I tried to move the `register_option(...)` to be located in the module that is concerned with the option, but that creates an issue when parsing the configuration: if the module in question is not imported yet, its configuration options are not registered yet and so the contents of the configuration cannot be validated.

This could be solved by:
* listing all options in the config module instead of in the various modules
* deferring validation of config contents to the point where the option values are used, hoping that the code that uses the option is in the same module that registered the option, so the option will have been registered by then (I don't like this because I think it would be better for the user if they get a crash asap if the config contains errors)
* importing all modules that declare options when the config module is loaded (not great because why import code that you won't use?)
* having in each module a `config` file that registers options for the module, and import all of these when the top-level `config` module is loaded. Eg. `otlLib.config`, `varLib.config` etc

@anthrotype or @behdad  would you be able to advise please?

image

**Comment by anthrotype** [comment]
The first option -- "listing all options in the config module instead of in the various modules" --  is the easiest, though probably not the most elegant.
The fourth option (otlLib.config, varLib.config etc) may also work, however in fonttools the subpackages `__init__.py` are usually not empty, so importing `varLib.config` would also import varLib's `__init__`.

What about, instead of needing to have calls to `register_option()` from some python module, you define a text file or files (e.g. configparser's .ini files) that contain the names of the modules and the options and the respective default values, and upon importing the global `fontTools.config` we parse that file or files (fetching it via importlib.resources and embedded in the distribution as package_data) and populate the Config class (calling register_option in turn for each one defined in the default config)?

**Comment by belluzj** [comment]
> a text file or files (e.g. configparser's .ini files) that contain the names of the modules and the options and the respective default values

That sounds OK except it's not Python code, so one would not be able to implement parsing/validation logic (or one would have to express that declaratively, like with JSON Schema for example). For that reason I would say it would be better as Python code. Also, I was quite happy with the idea that you could import the config options into your code, to refer to them without having hardcoded strings in the code; that wouldn't be simple with the text-file-approach.

Maybe the first option can be made to look organized enough by having `config.varLib`, `config.otlLib` etc modules to store the options of each module? It's still the same as option 1 but maybe it would make immediate sense to contributors. What do you think about that?

**Comment by anthrotype** [comment]
> implement parsing/validation logic

**Comment by anthrotype** [comment]
> implement parsing/validation logic

can you make an example of what you have in mind? I thought the options values would be simple str, float, int, or stuff that you can store as json.

> you could import the config options into your code, to refer to them without having hardcoded strings in the code

Not sure what you mean by hard coded string. You could still import the default config object and get the already-parsed values from it.

**Comment by anthrotype** [comment]
> config.varLib, config.otlLib etc modules to store the options of each module?

yeah, if you actually need to have _code_ to register options, then that could be an option to keep it organized. I'm still not 100 % that a declarative approach wouldn't be sufficient for our use case.

**Comment by belluzj** [comment]
> can you make an example of what you have in mind? I thought the options values would be simple str, float, int, or stuff that you can store as json.

Yes it's indeed very simple stuff, but still, assuming people can give options by writing them in a config file, we should expect to get the wrong type or values out of range. JSON schema would handle these cases for sure. See the tests below: parsing an input string to int, checking that the value is in the expected range.

https://github.com/fonttools/fonttools/blob/bd0e1563ead0f309237cff15800e8a169caf63fb/Tests/config_test.py#L44-L58

> Not sure what you mean by hard coded string. You could still import the default config object and get the already-parsed values from it.

See the two tests below: you have the `config` object, and you can index it either by string or by the `Option` object. I feel that indexing by option object should be the preferred way, it would help prevent typos; but I admit it's not a massive win at all, just a small thing.

https://github.com/fonttools/fonttools/blob/bd0e1563ead0f309237cff15800e8a169caf63fb/Tests/config_test.py#L20-L30

**Comment by behdad** [comment]
Collecting options from source into a toplevel config.py module is another option. We can make that run in CI and fail CI if config.py changed.

**Comment by behdad** [comment]
I'll comment on the rest of the approach soon. Thanks!

**Comment by belluzj** [comment]
Cosimo: don't perform conversions from string to int in the default constructor; have a special one for parsing from file that does the conversions. In general you wouldn't expect an object that accepts an int value to also accept a string value and convert it.

File: Lib/fontTools/otlLib/optimize/gpos.py [hunk 1]
```diff
@@ -3,22 +3,42 @@
...
-
-# NOTE: activating this optimization via the environment variable is
-# experimental and may not be supported once an alternative mechanism
-# is in place. See: https://github.com/fonttools/fonttools/issues/2349
-GPOS_COMPACT_MODE_ENV_KEY = "FONTTOOLS_GPOS_COMPACT_MODE"
-GPOS_COMPACT_MODE_DEFAULT = "0"
+from fontTools.config import register_option
+
+MODE_OPTION = register_option(
+    name="otlLib.optimize.gpos.mode",
```
Comment (belluzj) [inline Lib/fontTools/otlLib/optimize/gpos.py:15]: Cosimo: prepend `fontTools.`, actually use `f'{__name__}.mode'` (but that only works if the option is defined in the module)

**Comment by belluzj** [comment]
Instead of reading configuration from files, we could read it from a lib key in the designspace / UFO or a custom parameter in Glyphs.app source. 

E.g. in designspace:
```xml
<lib>
  <key>org.fontTools.config</key>
  <dict>
    <key>fontTools.otlLib.optimize.gpos.mode</key>
    <int>8</int>
  </dict>
</lib>
```

in ufo2ft: 

```python
ttFont = TTFont(config=designspace.lib.get('org.fontTools.config', {}))
```

**Comment by belluzj** [comment]
Cosimo: should this module be in misc? It's not font-related

Also maybe there's an off-the-shelf thing in general for all this?

**Comment by behdad** [comment]
> Instead of reading configuration from files, we could read it from a lib key in the designspace / UFO or a custom parameter in Glyphs.app source.

It can come from there as well, yes.



> ```
>     <key>fontTools.otlLib.optimize.gpos.mode</key>
> ```

I find `mode` very confusing. Can we call this `compressionLevel`?

**Comment by behdad** [comment]
I feel like the module is fundamental enough that should be at toplevel `fonttools.config`. But I'm fine if it goes in `misc.config`, and for everyone to use. That's a pattern we've been following. And I don't see anything simple off the shelf that we can use.

As for the TTFont property, at the risk of bikeshedding, I suggest we call it `font.cfg`. :D.

File: Lib/fontTools/varLib/plot.py [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 from fontTools.varLib.models import VariationModel, supportScalar
 from fontTools.designspaceLib import DesignSpaceDocument
-from matplotlib import pyplot
+# from matplotlib import pyplot
```
Comment (belluzj) [inline Lib/fontTools/varLib/plot.py:5]: TODO: undo this change

File: Lib/fontTools/otlLib/optimize/gpos.py (hunk 1 above)
Comment (belluzj) [reply Lib/fontTools/otlLib/optimize/gpos.py]: @anthrotype I did this but I went with the simple option 1, so I'm not using `__name__`

**Comment by belluzj** [comment]
* about where to define options: I implemented the simplest solution (all options to be listed in the config module's `__init__.py`)
* the module stays top-level but it's split between `fontTools.config.generic` (not related to fonts) and `fontTools.config` (the one to use everywhere in fontTools and client libraries). I could move the `generic` part to `misc` if that makes sense? Or not.
* I renamed `TTFont.config` to `TTFont.cfg` and `mode` to `compressionLevel`
* I'm using a dataclass in the config module and rebased on top of `drop-3.6` so that tests pass (sorry for the noise in the diffs, I'll rebase on top of main when possible)

**Comment by belluzj** [comment]
@anthrotype @behdad if you agree with how it is at the moment I'll start adding other options. I would also be happy to know which options you have in mind to add; I'm think to start with the parameters to TTFont().

File: Lib/fontTools/config/generic.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                value = option.parse(value)
+            except Exception as e:
+                raise ConfigValueParsingError(name, value) from e
+
+        if not option.validate(value):
+            raise ConfigValueValidationError(name, value)
+
+        self._values[name] = value
+
+    def get(self, option: Union[Option, str], default=_USE_GLOBAL_DEFAULT):
```
Comment (anthrotype) [inline Lib/fontTools/config/generic.py:139]: what's the point of allowing either an Option or a str for the key parameter of .get() and other dict-like methods? you're only using the Option's `name` attribute, a str, to lookup the global OPTIONS dict, so I think just accepting str would be enough, no?
What's the rationale?

**Comment by behdad** [comment]
Thanks Jany. This is concise and good. I have some ideas forward. Let me sketch those.

File: Lib/fontTools/config/__init__.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+Define all configuration options that can affect the working of fontTools
+modules. E.g. optimization levels of varLib IUP, otlLib GPOS compaction, etc.
+If this file gets too big, split it into smaller files per-module.
+
+An instance of the Config class can be attached to a TTFont object, so that
+the various modules can access their configuration options from it.
+"""
+from textwrap import dedent
+
+from fontTools.config.generic import *
```
[the remark this answers is not in the document]
Comment (belluzj) [reply Lib/fontTools/config/__init__.py:11]: Yes that makes sense, will do! I forgot about the __init__ behaviour you mention.

File: Lib/fontTools/otlLib/optimize/gpos.py [hunk 5]
```diff
@@ -3,40 +3,16 @@
...
-    ),
-    default=0,
-    parse=int,
-    validate=lambda v: v in range(10),
-)
+from fontTools.config import OPTIONS
 
 log = logging.getLogger("fontTools.otlLib.optimize.gpos")
 
+LEVEL_OPTION = OPTIONS['fontTools.otlLib.optimize.gpos.compressionLevel']
```
[the remark this answers is not in the document]
Comment (belluzj) [reply Lib/fontTools/otlLib/optimize/gpos.py]: Will do both. The colon sounds a bit strange to me at first glance but maybe it will make lots of sense once I see it in the code.

File: Lib/fontTools/config/generic.py (hunk 3 above)
Comment (belluzj) [reply Lib/fontTools/config/generic.py:139]: My idea here was that it is (marginally) more convenient and prevents typos. Imagine with just strings:

```python
def function():
    font.cfg['long option name without typo']
    ...lots of code...
    if something:
        font.cfg['long option name with a typo']
    ...lots of code...
    font.cfg['long option name without typo']
```
* long strings everywhere
* no editor autocomplete when using options/no squiggly underlines from the linter
* the typo would be silent until the code reaches the line of code with the typo (hopefully there's a unit test that covers it)

with the PR as it is now:

```python
THING = OPTIONS['long option name without typo']

def function():
    font.cfg[THING]
    ...lots of code...
    if something:
        font.cfg[THING]
    ...lots of code...
    font.cfg[THING]
```

* long string only once at the top
* autocomplete + linter to pick typos in the constant name
* if there's a typo in the long string it will explode on import

I do agree that these are subjective and marginal benefits, I'm happy to simplify the code and just use strings.

Comment (anthrotype) [reply Lib/fontTools/config/generic.py:139]: well, one could still do `font.cfg[THING.name]`.

**Comment by behdad** [comment]
Copy/pasting from chat to Jany/Cosimo:

Jany, Cosimo, I have some Evil config ideass 😀
gimme a week to cook them.
Here's the sneak peak:
based on your design so far.
In your __getitem__
```
def __getitem__(s):

  if hasattr(s, '__module__'):
    s = s.__module__ + s.__name__
  elif hasattr(s, '__name__'):
    s = s.__name__
```
then accept a tuple: (path,node):
```
  s = path+node
```
in short, you can then do:
```
  cfg[obj,"COMPRESSION_LEVEL"]
```
and avoid repeating `"fontTools.ttLib....."`

any classtype or module can be passed there.

better yet, instead of a tuple, we can accept a colon as Cosimo liked 😛
```
  cfg[obj:"COMPRESSION_LEVEL"]
```

Next. If only a path is passed and no node, we return the subconfig. That is, `cfg[cls]` will return a new cfg that has subconfig for the class. `__getattr__` will also be overriden to do the right thing.  So we can do: `cfg[cls].COMPRESSION_LEVEL`

Finally, `__call__` will be used to set config on multiple objects as a context:

```
with cfg(obj1, obj2, obj3, mod1, cls1, ...):
  ...
```
and the main way to use the config is to use to populate key/value arguments of constructors, methods, and functions. The syntax for that I have not fully figured out yet.
```
  cfg.invoke(obj.method) (params)
```
That would pass all relevant cfg to method.

Anyway; this is what's been brewing in my head.

As for your Option type, I want use the typing module types directly.

No need for extra Option type.

Behdad Esfahbod,
Lemme copy/paste this on the issue.

Jany Belluz,
Ok, thanks

That sounds a bit complicated to be honest, but maybe it looks nice when using it in the code

Behdad Esfahbod,
Yeah I feel like ergonomically it works.

Jany Belluz,
Ok cool

(sounds like how a Ruby developer would think, not a Python one)
I was going for pedestrian and non-surprising

Behdad Esfahbod,
I understand 🙂

I'm not sure mine are good. Evil it might be.

But want to explore 😀

Jany Belluz,
Yes anyway on the issue would be good 🙂 I'm curious to see how the usage looks

**Comment by belluzj** [comment]
@anthrotype I rebased on latest main

File: Lib/fontTools/misc/configTools.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        raise ConfigAlreadyRegisteredError(name)
+    option = Option(name, help, default, parse, validate)
+    OPTIONS[name] = option
+    return option
+
+
+_USE_GLOBAL_DEFAULT = object()
+
+
+class Config(MutableMapping):
```
Comment (anthrotype) [inline Lib/fontTools/misc/configTools.py:95]: I suggest we give Config a nice `__repr__` to aid debugging, I think MutableMapping doesn't make one for you.

File: Lib/fontTools/misc/configTools.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        )
+
+
+@dataclass
+class Option:
+    name: str
+    help: str
+    default: Any
+    # Turn input (e.g. string) into proper type. Only when reading from file.
+    parse: Callable
```
Comment (anthrotype) [inline Lib/fontTools/misc/configTools.py:74]: Callable[[str], Any]

File: Lib/fontTools/misc/configTools.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    # Return true if the given value is an acceptable value
+    validate: Callable
```
Comment (anthrotype) [inline Lib/fontTools/misc/configTools.py:76]: Callable[[Any], bool]

File: Lib/fontTools/misc/configTools.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+            self.set(name, value, parse_values)
+
+    def set(self, name: str, value: Any, parse_values=False):
+        """
+        Set the value of an option.
+        """
+        try:
+            option = OPTIONS[name]
+        except KeyError:
+            # FIXME: (Jany)  What to do with options we don't recognize?
```
Comment (anthrotype) [inline Lib/fontTools/misc/configTools.py:115]: could be an init parameter to the Config class, like `skip_unknown=False`

File: Lib/fontTools/misc/configTools.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+    def __init__(self, name, value):
+        super().__init__(
+            f"Config option {name} is unknown (with given value {repr(value)})"
+        )
+
+
+@dataclass
+class Option:
+    name: str
```
Comment (belluzj) [inline Lib/fontTools/misc/configTools.py:70]: Cosimo: delete this attribute, not needed because it's already in the dictionary key of OPTIONS

File: Lib/fontTools/misc/configTools.py (hunk 6 above)
Comment (belluzj) [reply Lib/fontTools/misc/configTools.py:95]: Print `f"Config({repr(self._values)})"`

File: Lib/fontTools/misc/configTools.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+    help: str
+    default: Any
+    # Turn input (e.g. string) into proper type. Only when reading from file.
+    parse: Callable
+    # Return true if the given value is an acceptable value
+    validate: Callable
+
+
+OPTIONS: Dict[str, Option] = {}
```
Comment (belluzj) [inline Lib/fontTools/misc/configTools.py:79]: Make object with Mapping interface, have register_option as method, move global variable to fontTools.config.__init__ (and have the inner dict of options double underscore prefixed)

File: Lib/fontTools/misc/configTools.py (hunk 9 above)
Comment (belluzj) [reply Lib/fontTools/misc/configTools.py:115]: Cosimo says: Or maybe a version the config file? Needs more thinking (maybe see what other applications are doing with their config). Maybe set() should always raise if you try to set an unknown option, but when you're parsing a file, maybe there it would make sense to skip unknown options?

File: Lib/fontTools/misc/configTools.py [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+
+_USE_GLOBAL_DEFAULT = object()
+
+
+class Config(MutableMapping):
+    _values: Dict
+
+    @classmethod
+    def from_file(cls):
+        """TODO"""
```
[the remark this answers is not in the document]
Comment (belluzj) [reply Lib/fontTools/misc/configTools.py]: I removed the method for now

File: Lib/fontTools/varLib/merger.py [hunk 13]
```diff
@@ -850,10 +846,10 @@ def merge(merger, self, lst):
 		# Compact the merged subtables
 		# This is a good moment to do it because the compaction should create
 		# smaller subtables, which may prevent overflows from happening.
-		mode = os.environ.get(GPOS_COMPACT_MODE_ENV_KEY, GPOS_COMPACT_MODE_DEFAULT)
-		if mode and mode != "0":
+		level = merger.font.cfg[COMPRESSION_LEVEL.name]
```
[the remark this answers is not in the document]
Comment (belluzj) [reply Lib/fontTools/varLib/merger.py]: I removed that and used the string directly everywhere

File: Lib/fontTools/misc/configTools.py (hunk 9 above)
Comment (belluzj) [reply Lib/fontTools/misc/configTools.py]: I added skip_unknown=False

**Comment by belluzj** [comment]
@anthrotype I fixed all your comments and added the new module to the Sphinx documentation.

I had an idea: how neat would it be if the list of registered options was parsed by Sphinx and added to the documentation page as well? I had a quick look and it sounds possible by adding some Sphinx code with new nodes and stuff, but I didn't go down the rabbit hole. Just a thought for later :)

**Comment by anthrotype** [comment]
> would it be if the list of registered options was parsed by Sphinx 

that would be nice, yeah. File an issue so we don't forget

**Comment by belluzj** [comment]
Done for the classvar and the issue

File: Lib/fontTools/ttLib/ttFont.py [hunk 14]
```diff
@@ -101,6 +102,7 @@ def __init__(self, file=None, res_name_or_index=None,
 		self.recalcTimestamp = recalcTimestamp
 		self.tables = {}
 		self.reader = None
+		self.cfg = Config(cfg)
```
Comment (anthrotype) [inline Lib/fontTools/ttLib/ttFont.py:105]: how about here you check `if not isinstance(cfg, AbstractConfig): cfg = Config(cfg)` otherwise you self `self.cfg = cfg`. This would allow one to stick a custom super-set of fonttools own Config with additional flags that can be useful in ones workflow

Comment (belluzj) [reply Lib/fontTools/ttLib/ttFont.py:105]: Ok will do now

Comment (anthrotype) [reply Lib/fontTools/ttLib/ttFont.py:105]: We would still want each TTFont instance to have its distinct copy, since a config is mutable.
You current cast `Config(cfg)` achieves that already. To accomplish the same with my proposal above we need a `AbstractConfig.copy()` method that returns a new instance of the type with its own copy of the internal `self._values` dictionary. A shallow copy is fine, if any mutable values are used in individual configs, it's not our job to deepcopy those I think.

So in addition to [my earlier suggestion](https://github.com/fonttools/fonttools/pull/2416/files#r850336161), we add this to `AbstractConfig`:

```python
def copy(self):
    return self.__class__(self._values)
```

and then call it inside `TTFont.__init__`

```python
    self.cfg = Config(cfg) if not isinstance(cfg, AbstractConfig) else cfg.copy()
```

Comment (anthrotype) [reply Lib/fontTools/ttLib/ttFont.py:105]: maybe also add a test that sets identical Config oject in two different fonts, and assert they have different `id()`;  or modify one font.cfg, and assert that it did not also change in the other font

Comment (belluzj) [reply Lib/fontTools/ttLib/ttFont.py]: Done, I added a test for the use-case that you were describing, if you can check that it's what you meant?

Comment (anthrotype) [reply Lib/fontTools/ttLib/ttFont.py]: > if you can check that it's what you meant?

yes, that's it

Comment (belluzj) [reply Lib/fontTools/ttLib/ttFont.py]: Done for real now including the copy()

File: Lib/fontTools/ttLib/ttFont.py [hunk 15]
```diff
@@ -89,7 +91,7 @@ def __init__(self, file=None, res_name_or_index=None,
 			sfntVersion="\000\001\000\000", flavor=None, checkChecksums=0,
 			verbose=None, recalcBBoxes=True, allowVID=NotImplemented, ignoreDecompileErrors=False,
 			recalcTimestamp=True, fontNumber=-1, lazy=None, quiet=None,
-			_tableCache=None):
+			_tableCache=None, cfg={}):
```
Comment (anthrotype) [inline Lib/fontTools/ttLib/ttFont.py:94]: in general mutable default arguments are bad, but in this case you always make a copy so it should be fine. Alternatively you default to `cfg=None` and then in method body do:

```
if cfg is None:
    cfg = Config()
self.cfg = cfg.copy() if isinstance(cfg, AbstractConfig) else Config(cfg)
```

Comment (anthrotype) [reply Lib/fontTools/ttLib/ttFont.py:94]: actually it's fine like this

**Review by anthrotype — APPROVED** [review]

**Review by anthrotype — APPROVED** [review]

**Comment by belluzj** [comment]
@anthrotype I wrote release notes in the description
